accepted/2020/workloads/workload-resolvers.md

```diff
@@ -0,0 +1,237 @@
+  - Workload manifest packages should be laid out in folders under the root corresponding to the name of the workload manifest, but should not include a version in the path.  They should be updated in-place.  This allows the SDK to import all of the active `WorkloadManifest.targets` files with a simple wildcard import
+- Include the following .targets import:
+
+```xml
+<Import Project="$(WorkloadManifestRoot)/*/WorkloadManifest.targets" />
+```
+- Run target to generate appropriate build error if there are any `MissingWorkloadPack` items
+    - This should run early in the build, probably by hooking to run before the `_CheckForInvalidConfigurationAndPlatform` target
+    - The target should (via a task) look at the workload manifests to determine which workload need to be installed to supply the workload packs which are missing.  The error message generated should include this list of workloads that need to be installed.
+
+- Include the following import in one of its .props files:
+
+```xml
+<Import Project="AutoImport.props" Sdk="Microsoft.NET.SDK.WorkloadAutoImportPropsLocator">
+```
+
+### VS behavior
+
+When the VS project system loads a project, it should check the evaluation result for `MissingWorkloadPack` items.  If there are any, then it should raise the VS event to veto the project load and trigger the in-product acquisition experience for those workloads.
+
+In the same way that the SDK error generation does, the project system should map from the missing workload packs to the missing workloads via the workload manifests.  Additionally, it will need to map from the workload IDs that the .NET SDK uses to the right IDs that VS will use to install the workloads.  That mapping should either be defined in the workload manifest or in another file that is part of the same packgae.
+
+QUESTION: Will it be OK to put this mapping of .NET workload IDs to VS workload/component IDs in the workload manifest packages?  Or do we need to supply it some other way?
+
+### `dotnet bundle restore` (query project and references for workloads that are required)
+
+When we add support for `dotnet bundle restore`, we will need to be able to run a target that will gather all of the workloads required by a project and its project references.  We will implement this via a new entry point target that walks the project graph and collects the `MissingWorkloadPack` items from each one.
+
+## Other experience improvements
+
+The additional features we plan to add to MSBuild SDK resolvers will also enable us to make the following improvements in experiences that are not directly related to workloads.
+
+## NuGet SDK Resolver
+
+MSBuild includes a [NuGet-based MSBuild SDK resolver](https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-use-project-sdk?view=vs-2019#how-project-sdks-are-resolved).  It will be used if a version number is specified for the SDK, either in the SDK attribute or in a global.json file.
```

Comment:
> It will be used if a version number is specified for the SDK, either in the SDK attribute or in a global.json file.

I don't quite get this (I'm probably just missing context).  I was expecting this to say "If you had an sdk reference other than MS.Net.Sdk".

Reply:
You have to specify a version number in order for the NuGet SDK resolver to kick in.  You can specify that either directly in the Sdk attribute: `<Project Sdk="MSBuild.Sdk.Extras/2.0.54">`, or in global.json:

```json
{
  "msbuild-sdks": {
    "MSBuild.Sdk.Extras": "2.0.54"
  }
}
```

Reply:
I see. That seems like overly precise wording.

```suggestion
MSBuild includes a [NuGet-based MSBuild SDK resolver](https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-use-project-sdk?view=vs-2019#how-project-sdks-are-resolved).  It will be used if a custom SDK is specified, as an SDK name and version number, either in the SDK attribute in the project file or in a global.json file.
```